Yoast/javascript
# P2-252 - Date instruction for Schema blocks package

## Summary


This PR can be summarized in the following changelog entry:

* [schema-blocks] Adds a Date block instruction for adding dates to a schema block template.

## Relevant technical choices:

* Did not end up creating a Schema instruction, since the Attribute schema instruction can be used instead.
	* The date in the Schema is a string, saved as an attribute on the block. Since the saved attribute is already in ISO-8601 format, adding a new instruction does not add anything.
* I could not get a unit test to work that would test the `edit` method of the instruction.
	* After discussing this with the team we decided to skip the test for now and [solve it later](https://yoast.atlassian.net/browse/P2-603).

## Test instructions

This PR can be tested by following these steps:

*Note*: This PR has [a companion PR in Free](https://github.com/Yoast/wordpress-seo/pull/16543).

### Setting up everything for the test
* Checkout the `P2-252-Create-date-instruction` branch on `wordpress-seo` ([see this PR](https://github.com/Yoast/wordpress-seo/pull/16543)).
* Add a date instruction to the recipe **block** template:
	* Sample date instruction for adding a `publishedDate` date attribute:
	  ```
	  {{date name="publishedDate" }} 
	  ```
	* In the end, the block template should look like this:
	  ```php
	  <?php // phpcs:ignore Internal.NoCodeFound ?>
	  {{block name="yoast/recipe" title="Recipe" category="common" }}
	  {{sidebar-duration name="cook-time" output=false label="Cook time" }}
	  {{sidebar-duration name="prep-time" output=false label="Preparation time" }}
	  {{sidebar-input name="yield" output=false type="number" label="Serves #" }}
	  <div class={{class-name}}>
		  {{variable-tag-rich-text name="title" tags=[ "h1", "h2", "h3", "h4", "h5", "h6", "strong" ] placeholder="Recipe title" }}
		  {{inner-blocks required-blocks=[ { "name": "core/paragraph", "option": "Multiple" } , { "name": "core/image", "option": "One" } ] recommended-blocks=[ "core/image", "core/paragraph" ] appender="button" appenderLabel="Add to recipe" }}
    	  {{date name="publishedDate" }}
	  </div>
	  ```
* Add an attribute instruction to the recipe **schema** template:
	* Sample instruction:
	  ```
	  "publishedDate": {{attribute name="publishedDate" }}
	  ```
	* In the end, the schema template should look like this:
	  ```php
	  <?php // phpcs:ignore Internal.NoCodeFound ?>
	  {{schema name="yoast/recipe"}}
	  {
		  "@type": "Recipe",
		  "@id": "{{block-id}}",
		  "mainEntityOfPage": {
			  "@id": "%%main_schema_id%%"
		  },
		  "image": {{inner-blocks-id allowed-blocks=[ "core/image" ] only-first=true null-when-empty=true }},
		  "name": {{html name="title"}},
		  "author": {
			  "@id": "%%author_id%%"
		  },
		  "description": {{inner-blocks-html blocks={ "core/paragraph": "content" } null-when-empty=true allowed-tags=[ "h1","h2","h3","h4","h5","h6","br","a","p","b","strong","i","em" ] }},
		  "cookTime": {{attribute name="cook-time" }},
		  "prepTime": {{attribute name="prep-time" }},
		  "recipeInstructions": {{inner-blocks allowed-blocks=[ "yoast/steps" ] only-first=true }},
		  "recipeIngredient": {{inner-blocks allowed-blocks=[ "yoast/ingredients" ] only-first=true }},
		  "recipeYield": {{attribute name="yield" }}, // <--- Comma here!
    	  "publishedDate": {{attribute name="publishedDate" }}
	  }
	  ```
* Build the JavaScript files, in the `schema-blocks` package as well as the Free plugin.
	* E.g. `yarn build` in the `schema-blocks` package and `grunt build:js` in the Free plugin. 

### Testing in the editor
* Create a new post.
* Add a recipe block.
* You should see a gray text with the date format (e.g. `F j, Y`) as currently configured in the site settings.
* Click on the gray text, you should see a date picker, and select a date.
* The gray text should change to the selected date.
* Save the post.

### Testing on the site itself.
* View the saved post with the recipe on the site itself.
* Check the Schema that is added to the page.
	* It should contain a graph piece for the recipe, containing the added `publishedDate` attribute.
	* This attribute's value should be the date selected in the editor, but in ISO-8601 format (e.g. `YYYY-MM-DD`).
	* Example:
	  ```js
	  {
          "@type": "Recipe",
          "@id": "http://basic.wordpress.test/sushi/#/schema/yoast-recipe-0",
          "mainEntityOfPage": {
              "@id": "http://basic.wordpress.test/sushi/#webpage"
          },
          "name": "Sushi",
          "author": {
              "@id": "http://basic.wordpress.test/#/schema/person/a00dc884baa6bd52ebacc06cfd5aab21"
          },
          "recipeInstructions": [
              {
                  "@type": "HowToStep",
                  "name": "Cook the rice.",
                  "url": "http://basic.wordpress.test/sushi/#/schema/yoast-step-0-1-0"
              }
          ],
          "recipeIngredient": [
              "Rice",
              "Salmon",
              "Seaweed"
          ],
          "publishedDate": "2021-01-21" // <-- This.
      }
	  ```

### Other date formatting
* Go to your site settings.
* Change your date format, e.g. to `Y-M-d`.
* Create a new post.
* Add a recipe block.
* Test the date instruction like before, on the editor as well as on the site itself, but see that the date format has changed to be the one configured in the site's settings.

## Impact check

* This PR affects the following parts of the plugin, which may require extra testing:
  * N/A, since this functionality is behind the Schema blocks feature flag.

## UI changes

## Quality assurance


Fixes #
